proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+<!-- {% raw %} -->
+
+# Out-of-process compiler api architecture
+
+* Proposal: [NNNN](NNNN-outofproc-compiler-api-architecture.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose a more detailed architecture for the approved 
+proposal: [0005](0005-inproc-outofproc-compiler-api-support.md).
+
+The c api [0008](0008-c-interface-compiler-library.md) for compiling a shader
+will be implemented using this out of process architecture.
+
+## Proposed solution
+
+The architecture for an out of process design will take advantage of systems
+that have multiple processors, or multiple-core processors. A process pool will
+be created that creates a separate compiler process for each available
+processor. For example, if the system has four processors, then four compiler
+processes are created.
+
+The process pool is owned by a singleton object and shared between each api
+instance created in the calling process.
```

Comment:
"singleton object" is a bit of a implementation detail that I'm not sure belongs here.

```suggestion
Each process that uses the API has a single process pool that is shared between all instances of the API that process creates.,
```

Reply:
Possible alternative: there's a machine-global process pool that is shared between all processes that use the API.

(Might be worth including in the Alternatives Considered section)

Reply:
I removed wording around the singleton implementation detail.  I will add your possible alternative suggestion as it is worth a conversation at least.